### ISSUE TITLE
Ugly fail when manifests files differ

### ISSUE THREAD

**Issue by lasote** [conan-io/conan#1040]
Initially reported by @claasd in #1033. Repair the ugly crash checking the key existence.

```
D:\slave\ws\cab\extern\boost\1.63.0@0\ws>conan.exe upload "Boost/1.63.0@cab/extern" --all --confirm -r bop --retry 3 
Uploading Boost/1.63.0@cab/extern

Compressing recipe...                                                 
Uploading conanmanifest.txt                                           


Uploading conan_export.tgz                                            

Uploaded conan recipe 'Boost/1.63.0@cab/extern' to 'bop': http://conan.bop
Uploading package 1/5: 4f0fd3886115d238859ac2a3d41664e4236efc61

Checking package integrity...                                         
Package integrity OK!                                                 

Compressing package...                                                
Requesting upload permissions...                                      
Requesting upload permissions...Done!                                 

Uploading conan_package.tgz                                           

Uploading package 2/5: 5d34366328ff5fe8b9f93b912e5855ffb5fda596

Checking package integrity...                                         
Traceback (most recent call last):
  File "C:\Python34\Scripts\conan-script.py", line 9, in <module>
    load_entry_point('conan==0.19.2', 'console_scripts', 'conan')()
  File "C:\Python34\lib\site-packages\conans\conan.py", line 6, in run
    main(sys.argv[1:])
  File "C:\Python34\lib\site-packages\conans\client\command.py", line 1065, in main
    error = command.run(args)
  File "C:\Python34\lib\site-packages\conans\client\command.py", line 979, in run
    raise exc
  File "C:\Python34\lib\site-packages\conans\client\command.py", line 959, in run
    method(args[0][1:])
  File "C:\Python34\lib\site-packages\conans\client\command.py", line 826, in upload
    retry_wait=args.retry_wait)
  File "C:\Python34\lib\site-packages\conans\client\manager.py", line 540, in upload
    retry=retry, retry_wait=retry_wait)
  File "C:\Python34\lib\site-packages\conans\client\uploader.py", line 41, in upload_conan
    self._upload_conan(conan_ref, force, all_packages, retry, retry_wait)
  File "C:\Python34\lib\site-packages\conans\client\uploader.py", line 56, in _upload_conan
    retry, retry_wait)
  File "C:\Python34\lib\site-packages\conans\client\uploader.py", line 76, in upload_package
    self._remote_proxy.upload_package(package_ref, retry, retry_wait)
  File "C:\Python34\lib\site-packages\conans\client\proxy.py", line 295, in upload_package
    result = self._remote_manager.upload_package(package_ref, remote, retry, retry_wait)
  File "C:\Python34\lib\site-packages\conans\client\remote_manager.py", line 77, in upload_package
    if read_manifest.file_sums[fname] != expected_manifest.file_sums[fname]:
KeyError: 'include/boost/asio/ip/unicast.hpp'
```

**Comment by claasd** [comment]
There were some messages before:
```
Boost/1.63.0@cab/extern: Building package from source as defined by build_policy='missing'
Boost/1.63.0@cab/extern: Building your package in D:\slave\ws\cab\extern\boost\1.63.0@0\ws/.cn\7hj72g45\1
Boost/1.63.0@cab/extern: Configuring sources in D:\slave\ws\cab\extern\boost\1.63.0@0\ws/.cn\1ok30bb7\1
Boost/1.63.0@cab/extern: Downloading https://netassist.dl.sourceforge.net/project/boost/boost/1.63.0/boost_1_63_0.zip...

Unzipping 521.9 MB, this can take a while
Error extract boost_1_63_0/libs/geometry/doc/html/geometry/reference/spatial_indexes/boost__geometry__index__rtree/rtree_parameters_type_const____indexable_getter_const____value_equal_const____allocator_type_const___.html
Filename too long
Boost/1.63.0@cab/extern: Copying sources to build folder
Boost/1.63.0@cab/extern: WARN: Filename too long, file excluded: D:\slave\ws\cab\extern\boost\1.63.0@0\ws\.cn\7hj72g45\1\boost_1_63_0\libs\geometry\doc\doxy\doxygen_output\html_by_doxygen\classboost_1_1geometry_1_1promote__integral_3_01_t_00_01_promote_unsigned_to_unsigned_00_01_use_0d823691c6018c3854e3ced6bc93fbb6.html
Boost/1.63.0@cab/extern: WARN: Filename too long, file excluded: D:\slave\ws\cab\extern\boost\1.63.0@0\ws\.cn\7hj72g45\1\boost_1_63_0\libs\geometry\doc\doxy\doxygen_output\html_by_doxygen\classboost_1_1geometry_1_1promote__integral_3_01_t_00_01_promote_unsigned_to_unsigned_00_01_use_checked_integer_00_01false_01_4.html
Boost/1.63.0@cab/extern: WARN: Filename too long, file excluded: D:\slave\ws\cab\extern\boost\1.63.0@0\ws\.cn\7hj72g45\1\boost_1_63_0\libs\geometry\doc\doxy\doxygen_output\html_by_doxygen\classboost_1_1geometry_1_1strategy_1_1transform_1_1translate__transformer_3_01_calculation_type_00_012_00_012_01_4-members.html
Boost/1.63.0@cab/extern: WARN: Filename too long, file excluded: D:\slave\ws\cab\extern\boost\1.63.0@0\ws\.cn\7hj72g45\1\boost_1_63_0\libs\geometry\doc\doxy\doxygen_output\html_by_doxygen\classboost_1_1geometry_1_1strategy_1_1transform_1_1translate__transformer_3_01_calculation_type_00_013_00_013_01_4-members.html
```

and so on.

But there is nothing else about unicast.hpp in the log

**Comment by claasd** [comment]
For info, the file seems to be random, here is the log of an earlier build:
```
Checking package integrity...                                         
Traceback (most recent call last):
  File "C:\Python34\Scripts\conan-script.py", line 9, in <module>
    load_entry_point('conan==0.19.1', 'console_scripts', 'conan')()
  File "C:\Python34\lib\site-packages\conans\conan.py", line 6, in run
    main(sys.argv[1:])
  File "C:\Python34\lib\site-packages\conans\client\command.py", line 1065, in main
    error = command.run(args)
  File "C:\Python34\lib\site-packages\conans\client\command.py", line 979, in run
    raise exc
  File "C:\Python34\lib\site-packages\conans\client\command.py", line 959, in run
    method(args[0][1:])
  File "C:\Python34\lib\site-packages\conans\client\command.py", line 826, in upload
    retry_wait=args.retry_wait)
  File "C:\Python34\lib\site-packages\conans\client\manager.py", line 540, in upload
    retry=retry, retry_wait=retry_wait)
  File "C:\Python34\lib\site-packages\conans\client\uploader.py", line 41, in upload_conan
    self._upload_conan(conan_ref, force, all_packages, retry, retry_wait)
  File "C:\Python34\lib\site-packages\conans\client\uploader.py", line 56, in _upload_conan
    retry, retry_wait)
  File "C:\Python34\lib\site-packages\conans\client\uploader.py", line 76, in upload_package
    self._remote_proxy.upload_package(package_ref, retry, retry_wait)
  File "C:\Python34\lib\site-packages\conans\client\proxy.py", line 295, in upload_package
    result = self._remote_manager.upload_package(package_ref, remote, retry, retry_wait)
  File "C:\Python34\lib\site-packages\conans\client\remote_manager.py", line 77, in upload_package
    if read_manifest.file_sums[fname] != expected_manifest.file_sums[fname]:
KeyError: 'include/boost/units/is_quantity.hpp'
```

**Comment by memsharded** [comment]
Don't know where the error could come from. Could you please trying running from sources applying this code?:

# remote_manager.py, line 77, in upload_package

```python
        for fname in read_manifest.file_sums.keys():
                try:
                    if read_manifest.file_sums[fname] != expected_manifest.file_sums[fname]:
                        self._output.warn("Mismatched checksum for file %s (checksum: %s, expected: %s)" %
                                          (fname, read_manifest.file_sums[fname], expected_manifest.file_sums[fname]))
                except KeyError:
                    self._output.warn("The expected manifest doesn't contain %s" % fname)
                    self._output.warn("Expected manifest files: %s" % ", ".join(expected_manifest.file_sums.keys()))
```

The only reason for this to fail is that such file has been removed from the package. The manifest is computed by traversing the existing files when the package is created. Then at upload, it is checked, and what is happening here is that such file has dissapeared. Maybe the above code gives us some hint. Thanks!

**Comment by claasd** [comment]
I will try to reproduce it, but I cannot guarantee that the error will show up again. 
As I remember, it was produced when I was playing around with compression levels. 
E.g. build with level 4, try upload, server rejected as too large, set compression level to 9 and started the upload again.

**Comment by memsharded** [comment]
Reviewing the code I don't find how the compression level could affect this, but who knows. Ok, sounds good, try to leave the patch for a while, and lets see if it fails again (it will print a warning, but not raise Error). Another possibility, maybe more convenient is run without the patch and the moment it fails with error, zip the package folder and send it to us for inspection.

**Comment by memsharded** [comment]
Hi @claasd, I am having a look into this, to fix the ugly failure. 
But still the origin of this is unknown. Have you had luck in chasing it? Have it reproduce again? Feedback about the status of this would be very helpful, thanks!

**Comment by lasote** [comment]
It has been released in conan 0.22.0. If you finally find the origin of the problem please report it.
